examples/cmd/pool.go

```diff
@@ -0,0 +1,173 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/src-d/metadata-retrieval/github"
+)
+
+type DownloadersPool struct {
+	Size    int
+	pool    chan *DownloaderPoolItem
+	started bool
+	ended   bool
+	t0      time.Time
+	t1      time.Time
+	stats0  map[int]*downloaderStats
+	stats1  map[int]*downloaderStats
+}
+
+type DownloaderPoolItem struct {
+	id int
+	*github.Downloader
+}
+
+type DownloaderPoolStats struct {
+	Elapsed    time.Duration
+	RatesUsage map[int]*RateUsage
+}
+
+func (d *DownloaderPoolItem) stats() (*downloaderStats, error) {
+	rate, err := d.RateRemaining(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+
+	return &downloaderStats{
+		Rate: rate,
+		Time: time.Now(),
+	}, nil
+}
+
+type downloaderStats struct {
+	Rate int
+	Time time.Time
+}
+
+type downloaderBuilder = func(c *http.Client) (*github.Downloader, error)
+
+func NewDownloadersPool(clients []*http.Client, builder downloaderBuilder) (*DownloadersPool, error) {
+	ch := make(chan *DownloaderPoolItem, len(clients))
+
+	for i, c := range clients {
+		d, err := builder(c)
+		if err != nil {
+			return nil, err
+		}
+
+		ch <- &DownloaderPoolItem{id: i, Downloader: d}
+	}
+
+	return &DownloadersPool{
+		Size: len(clients),
+		pool: ch,
+	}, nil
+}
+
+func (dp *DownloadersPool) WithDownloader(f func(d *DownloaderPoolItem) error) error {
+	if !dp.started || dp.ended {
+		return fmt.Errorf("invalid state: started=%v, ended=%v",
+			dp.started, dp.ended)
+	}
+
+	item := <-dp.pool
+	defer func() {
+		dp.pool <- item
+	}()
+
+	return f(item)
+}
+
+func (dp *DownloadersPool) Begin() error {
+	if dp.started || dp.ended {
+		return fmt.Errorf("invalid state for `Begin()`: started=%v, ended=%v",
+			dp.started, dp.ended)
+	}
+
+	dp.started = true
+	dp.t0 = time.Now()
+	stats0, err := dp.stats()
+	if err != nil {
+		return err
+	}
+
+	dp.stats0 = stats0
+	return nil
+}
+
+func (dp *DownloadersPool) End() (*DownloaderPoolStats, error) {
+	if !dp.started || dp.ended {
+		return nil, fmt.Errorf("invalid state for `End()`: started=%v, ended=%v",
+			dp.started, dp.ended)
+	}
+
+	dp.t1 = time.Now()
+
+	defer func() {
+		dp.ended = true
```

Comment:
why in `defer`?

Reply:
Just to avoid writing it twice before returning. But in this case actually even setting it just after `dp.stats()` it's ok.